### PR TITLE
TYP: dtype specializations for np.ndarray in algos.pyi (part)

### PR DESCRIPTION
**Pull request by simonjayhawkins**
cc @jbrockmendel

### CODE REVIEW

**Comment by jbrockmendel** [comment]
neat!  ill have to go through and double-check the exact dtypes, but this looks really nice

**Comment by simonjayhawkins** [comment]
Thanks. The obvious place to start is the libraries then propagate these up.

I'm struggling to get mypy failures. static type checking checks type compatibility whereas the cython is specialized for specific dtypes.

**Comment by jbrockmendel** [comment]
i take it npt.NDArray doesnt allow for specifying ndim?

File: pandas/tseries/frequencies.py [hunk 1]
```diff
@@ -405,11 +427,19 @@ def _get_wom_rule(self) -> str | None:
...
+        weekdays = unique(self.index.weekday)  # type: ignore[union-attr]
         if len(weekdays) > 1:
             return None
 
-        week_of_months = unique((self.index.day - 1) // 7)
+        # error: Item "DatetimeIndex" of "Union[DatetimeIndex, TimedeltaIndex]" has no
+        # attribute "day"
+        # error: Item "TimedeltaIndex" of "Union[DatetimeIndex, TimedeltaIndex]" has no
+        # attribute "day"
+        week_of_months = unique((self.index.day - 1) // 7)  # type: ignore[union-attr]
```
Comment (jbrockmendel) [inline pandas/tseries/frequencies.py:442]: doesnt need to be this PR, but looks like the solution here is to pin down self.index as DatetimeIndex here

File: pandas/_libs/algos.pyi [hunk 2]
```diff
@@ -28,86 +32,87 @@ class NegInfinity:
...
+    np.uint32,
+    np.uint16,
+    np.uint8,
+]
 
 def validate_limit(nobs: int | None, limit=None) -> int: ...
 def pad(
-    old: np.ndarray,  # ndarray[algos_t]
-    new: np.ndarray,  # ndarray[algos_t]
+    old: npt.NDArray[algos_t],  # ndarray[algos_t]
```
Comment (jbrockmendel) [inline pandas/_libs/algos.pyi:85]: can get rid of the comments for these

**Comment by jbrockmendel** [comment]
needs rebase, otherwise i think this is an improvement even if you're having trouble producing errors

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
@simonjayhawkins what are your plans here?  i was thinking of doing a pass updating the _libs ndarray annotations

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins what are your plans here? i was thinking of doing a pass updating the _libs ndarray annotations

you are welcome to pull these changes into another branch.

(or I can fix-up this PR. I did not think it was appropriate to work on typing PRs while there are typing PRs from others that need review)

**Comment by jbrockmendel** [comment]
> I did not think it was appropriate to work on typing PRs while there are typing PRs from others that need review

thats thoughtful of you, but no one will hold it against you

**Pull request closed without merging**